source/common/http/conn_manager_impl.h

```diff
@@ -533,6 +538,37 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void callHighWatermarkCallbacks();
     void callLowWatermarkCallbacks();
 
+    void setRequestHeaders(RequestHeaderMapPtr&& request_headers) {
+      ASSERT(request_headers_ == nullptr);
+      request_headers_ = std::move(request_headers);
+    }
+
+    void setResponseHeaders(ResponseHeaderMapPtr&& response_headers) {
+      // Note: sometimes the headers get reset (local reply while response is buffering), so we
+      // don't assert here.
+      response_headers_ = std::move(response_headers);
+    }
```

Comment:
The existence of these functions seems odd to me. I already commented on response headers, but for request headers couldn't the filter manager store the request headers when it's decode headers method is called? Why does it have to be done out of band?

Reply:
It gets a bit tricky with local replies: several code paths in `decodeHeaders` might send a local reply before we'd dispatched the request header to the filter manager. I think the main reason why we need the headers set before we do a local reply is that we want to make sure that they appear in access logs.

I think one option would be to have a `sendLocalReply` call that accepts a `RequestHeaderMapPtr&&` that takes ownership over the headers before sending a local reply, with ASSERTs verifying that we never call the other `sendLocalReply` overload without headers haven't being set via `decodeHeaders`?

Reply:
> I think one option would be to have a sendLocalReply call that accepts a RequestHeaderMapPtr&& that takes ownership over the headers before sending a local reply, with ASSERTs verifying that we never call the other sendLocalReply overload without headers haven't being set via decodeHeaders?

Yeah this sounds good. Maybe just put in TODOs for now and work on cleaning this up later once the code is split?

/wait